File: bots/cpp/random/random.cpp
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <vector>
#include <string>

#include <boost/algorithm/string/join.hpp>
#include <boost/algorithm/string/replace.hpp>

#include <netorcai-client-cpp/client.hpp>
#include <netorcai-client-cpp/error.hpp>

#include "hexabomb-parse.hpp"

int main()
{
    srand(time(0));

    using namespace netorcai;
    using namespace std;

    try
    {
        Client c;

        printf("Connecting to netorcai... "); fflush(stdout);
        c.connect();
        printf("done\n");

        printf("Logging in as a player... "); fflush(stdout);
        c.sendLogin("C++-player", "player");
        c.readLoginAck();
        printf("done\n");

        std::unordered_map<Coordinates, Cell> cells;
        std::vector<Character> characters;
        std::vector<Bomb> bombs;
        std::map<int, int> score, cellCount;

        printf("Waiting for GAME_STARTS... "); fflush(stdout);
        const GameStartsMessage gameStarts = c.readGameStarts();
        parseGameState(gameStarts.initialGameState, cells, characters, bombs, score, cellCount);
        int myColor = gameStarts.playerID + 1;
        printf("done\n");

        const vector<string> possibleActions = {
            R"({"id":ID, "movement":"move", "direction":"x+"})",
            R"({"id":ID, "movement":"move", "direction":"y+"})",
            R"({"id":ID, "movement":"move", "direction":"z+"})",
            R"({"id":ID, "movement":"move", "direction":"x-"})",
            R"({"id":ID, "movement":"move", "direction":"y-"})",
            R"({"id":ID, "movement":"move", "direction":"z-"})",
            R"({"id":ID, "movement":"bomb", "bomb_delay":3, "bomb_range":3})",
            R"({"id":ID, "movement":"revive", "revive_q":0, "revive_r":0})",
        };

        for (int i = 1; i < gameStarts.nbTurnsMax; i++)
        {
            printf("Waiting for TURN... "); fflush(stdout);
            const TurnMessage turn = c.readTurn();
            parseGameState(turn.gameState, cells, characters, bombs, score, cellCount);
            printf("done\n");

            // Take a random action for each character of the bot.
            vector<string> actions;
            for (const auto & character : characters)
            {
                if (character.color == myColor)
                {
                    // Pick a random action.
                    string action = possibleActions[rand()%possibleActions.size()];

                    // Replace ID by the character ID. Yup, still not in C++ standard lib...
                    boost::replace_all(action, "ID", to_string(character.id));
                    actions.push_back(action);
                }
            }

            string mergedActions = "[" + boost::join(actions, ", ") + "]";
            printf("Sending actions=%s... ", mergedActions.c_str()); fflush(stdout);
            c.sendTurnAck(turn.turnNumber, json::parse(mergedActions));
            printf("done\n");
        }

        return 0;
    }
    catch (const netorcai::Error & e)
    {
        printf("Failure: %s\n", e.what());
        return 1;
    }
}

File: bots/java/src/bot/Random.java
package org.hexabomb.bot;

import java.util.*;
import org.json.*;
import org.netorcai.Client;
import org.netorcai.message.*;
import org.hexabomb.*;

public class Random
{
    public static String getRandomElement(ArrayList<String> arr)
    {
        java.util.Random rand = new java.util.Random();
        return arr.get(rand.nextInt(arr.size()));
    }

    public static void main(String [] args)
    {
        try
        {
            Client c = new Client();

            System.out.printf("Connecting to netorcai... ");
            System.out.flush();
            c.connect();
            System.out.println("done");

            System.out.printf("Logging in as a player... ");
            System.out.flush();
            c.sendLogin("java", "player");
            c.readLoginAck();
            System.out.println("done");

            System.out.printf("Waiting for GAME_STARTS... ");
            System.out.flush();
            GameStartsMessage gameStarts = c.readGameStarts();
            GameState initialGameState = GameState.parse(gameStarts.initialGameState);
            int myColor = gameStarts.playerID + 1;
            System.out.println("done");

            ArrayList<String> possibleActions = new ArrayList<String>();
            possibleActions.add("{\"id\":ID, \"movement\":\"move\", \"direction\":\"x+\"}");
            possibleActions.add("{\"id\":ID, \"movement\":\"move\", \"direction\":\"y+\"}");
            possibleActions.add("{\"id\":ID, \"movement\":\"move\", \"direction\":\"z+\"}");
            possibleActions.add("{\"id\":ID, \"movement\":\"move\", \"direction\":\"x-\"}");
            possibleActions.add("{\"id\":ID, \"movement\":\"move\", \"direction\":\"y-\"}");
            possibleActions.add("{\"id\":ID, \"movement\":\"move\", \"direction\":\"z-\"}");
            possibleActions.add("{\"id\":ID, \"movement\":\"bomb\", \"bomb_delay\":3, \"bomb_range\":3}");
            possibleActions.add("{\"id\":ID, \"movement\":\"revive\", \"revive_q\":0, \"revive_r\":0}");

            for (int i = 1; i < gameStarts.nbTurnsMax; i++)
            {
                System.out.printf("Waiting for TURN... ");
                System.out.flush();
                TurnMessage turn = c.readTurn();
                GameState gameState = GameState.parse(turn.gameState);
                System.out.println("done");

                // Take a random action for each character of the bot.
                ArrayList<String> actions = new ArrayList<String>();
                for (int j = 0; j < gameState.characters.size(); j++)
                {
                    org.hexabomb.Character character = gameState.characters.get(j);
                    if (character.color == myColor)
                    {
                        // Pick a random action.
                        String action = getRandomElement(possibleActions);

                        // Replace ID by the character ID.
                        action = action.replace("ID", String.valueOf(character.id));
                        actions.add(action);
                    }
                }

                String mergedActions = "[" + String.join(", ", actions) + "]";
                System.out.format("Sending actions=%s... ", mergedActions);
                c.sendTurnAck(turn.turnNumber,  new JSONArray(mergedActions));
                System.out.println("done");
            }
        }
        catch(Throwable t)
        {
            System.out.println(t);
        }
    }
}
